### DISCUSSION TITLE
Migrate from Intel (Rosetta2) to ARM brew on M1

### DISCUSSION THREAD

**Discussion by tyron** [Homebrew/discussions#417] [Getting started]
I noticed the progress on https://github.com/Homebrew/brew/issues/10152 and that the installation script has been updated to install under the new `/opt/homebrew` prefix on M1 Macs. I had brew installed through Rosetta2 (`/usr/local/homebrew`) and running `brew up && brew doctor` now tells me:
```
Warning: Your Homebrew's prefix is not /opt/homebrew.
```

What is the best way to move from Rosetta2 to ARM on M1 Macs? Should I follow uninstall instructions + new installation?
Running `brew upgrade` displays:
```
Error: Cannot install in Homebrew on ARM processor in Intel default prefix (/usr/local)!
Please create a new installation in /opt/homebrew using one of the
"Alternative Installs" from:
  https://docs.brew.sh/Installation
You can migrate your previously installed formula list with:
  brew bundle dump
```
But it's not clear what happens to the old installation... Does it get removed, do we need to handle 2 `brew` commands?

**Comment by carlocab** [comment]
The old installation stays there, allowing you to migrate it to the new `brew` installation at your convenience.

>do we need to handle 2 `brew` commands?

Unless you uninstall your Rosetta 2 installation, yes. Don't forget to add `/opt/homebrew/bin` to your `PATH`.

**Reply by derekschinke** [reply]
I'm unable to access my old installation now. `brew` in both Terminal.app and Terminal Rosetta.app don't show anything as being installed after running `brew bundle dump`

**Reply by carlocab** [reply]
If `/opt/homebrew/bin` appears in your `PATH` before `/usr/local/bin`, then calls to `brew` will always be to your M1 installation. You need to access your Intel installation with `/usr/local/bin/brew`.

**Reply by jchoi2x** [reply]
So do we run `rm -r /usr/local/Cellar` or something similar to clean up the old installation? Is there a reason for preserving the old installation once we migrated to the m1 version of brew?

**Reply by carlocab** [reply]
You can use the [uninstall script](https://github.com/Homebrew/install#uninstall-homebrew) to uninstall your old installation. You'll want to download it first and run `/bin/bash uninstall.sh --help` as suggested in order to know how to specify the prefix you want to uninstall from. If you don't specify this, it will assume you want to uninstall the one in `/opt/homebrew`.

>Is there a reason for preserving the old installation once we migrated to the m1 version of brew?

Some formulae don't work on ARM yet. The ones that do aren't all necessarily fully stable yet. So, you may wish to keep a Rosetta installation in the meantime.

**Reply by sumater** [reply]
So how do I purge the x86 versions of my ARM packages?

**Comment by JakobJingleheimer** [comment]
By default on Big Sur (or possibly due to migrating from High Sierra via Time Machine), `/opt/` is owned by user `root` and group `wheel` (root's group).

Is homebrew expecting a chown first? A bit of googling suggests there's no standard for who should own `/opt/`.

Relevance: the homebrew Alternative Installation instructions say to create a `/opt/homebrew` directory, which fails with `Permission denied!`.

**Reply by gromgit** [reply]
> Is homebrew expecting a chown first?

No. Homebrew on Apple Silicon expects `/opt/homebrew` to be owned by you, and the official installer ensures this, using `sudo` to Make It So if necessary.

AFAIK, there's only one requirement for `/opt` in Homebrew: its permissions/ownerships must not prevent you from reaching `/opt/homebrew` as a normal user. I don't own an AS box, but I also can't think of a reason for Apple engineers to lock down `/opt` to such an extent that `/opt/homebrew` becomes unreachable.

**Reply by JakobJingleheimer** [reply]
whilst running `brew bundle dump`, I encountered similar permission denied issues on `/opt/`. I temporarily changed the owner of `/opt/` to work around them, but I suspect there might be more in future (since I changed it back after).

also, bundle dump did not migrate any of my existing formulae (they are all still in `/usr/local/`). perhaps they are not compatible with ARM yet, but there were no errors / any output whatsoever (after resolving the permission denied issue).

**Reply by derekschinke** [reply]
same, `brew bundle dump` didn't transfer anything for me.

**Reply by SMillerDev** [reply]
No, it dumps the formula you installed to a brewfile so you can install that using the ARM based brew. It doesn't actually migrate anything.

**Reply by JakobJingleheimer** [reply]
That definitely wasn't clear…

thanks for the clarification though

**Reply by gromgit** [reply]
> whilst running `brew bundle dump`, I encountered similar permission denied issues on `/opt/`. I temporarily changed the owner of `/opt/` to work around them, but I suspect there might be more in future (since I changed it back after).

From the `brew` man page:
>       brew bundle dump
>           Write all installed casks/formulae/images/taps into a Brewfile in the current directory.

If you were in `/opt` when you ran `brew bundle dump`, that would explain your "permission denied" errors. If that wasn't the case, please post the _exact_ error you saw.

> also, bundle dump did not migrate any of my existing formulae (they are all still in `/usr/local/`).

The above extract from the man page explains why.

**Reply by JakobJingleheimer** [reply]
Yes, I was in `/opt/` at the time (as per the instructions: `mkdir homebrew`, which would need the cwd to be `/opt/`). I was following the error message from homebrew that instructed me to reinstall it and then run the bundle dump to migrate my existing formulae (which it doesn't).

I realise homebrew's target audience is tech-savvy, but it could definitely benefit from a bit more hand-holding rather than a shove out of the nest.

Did I follow the instructions homebrew outputted? Yes.

Did it address the problem / did homebrew start working? No.

Did I have any idea what I was supposed to do next? Also no.

After the vague "go find X and then do Y", none what I expected to happen did. I started googling, and the most helpful thing I found was a snarky comment in a bug that explaining why the new install was not picked up (something was required to be added to `.bash_profile`, which the instructions didn't mention nor did the installation output, as erroneously claimed by the snarky poster), and this post with @carlocab's helpful insight (which I vote be added to brew.sh's alternate installation instructions, cuz they're definitely needed).

**Comment by pragyasresta29** [comment]
This is what worked for me. If `/opt/homebrew` is before `/usr/local/bin/brew` then it uses the homebrew of M1 installation. If `/usr/local/bin/brew` is before `/opt/homebrew` then it uses the rosetta installation. 
My path in `~/.zshrc` looks something like this:
`export PATH=$HOME/bin:/opt/homebrew/bin:/usr/local/bin:$PATH`
So, now when I use brew, it uses the M1 installation and doesn't show the error. 
But the error now shows when I want to use the rosetta installation. I figured I just need to change the precedence of the path to use either one. For now, I use the homebrew for M1.

**Reply by NateLevin1** [reply]
You just saved my day. TYSM!

**Reply by avxkim** [reply]
is it safe to remove old brew installation at /usr/local?

**Reply by SMillerDev** [reply]
If you don't need anything from it, yes

**Comment by SaschaKo77** [comment]
What I did is running `brew bundle dump` as suggested by `brew upgrade`. This creates a file named Brewfile in the current directory. Then I installed the new version of brew for M1 in /opt/homebrew/. Finally adding  /opt/homebrew/bin to the path (at the beginning) and running `brew bundle --file Brewfile` installed everything in the new brew environment.

**Reply by lilyharuna** [reply]
You just saved my life. You're a genius! Thank you a lot. I resolved my problems. :)

**Reply by sumater** [reply]
Unfortunately I uninstalled Homebrew before doing `bundle dump`, so now I'm trying to find a way to migrate my x86 packages after the fact.

**Reply by gromgit** [reply]
@sumater Assuming:
* you've installed ARM Homebrew (i.e. `which brew` outputs `/opt/homebrew/bin/brew`, and `brew doctor` doesn't output any issues), and
* `/usr/local/Cellar` still contains the packages that you installed under Intel Homebrew,

then the sane way is to manually install only the packages you see in `/usr/local/Cellar` that you really need, _when you need them_. You'll likely find that you actually don't need everything you previously installed, which is a Good Thing.

Others will likely suggest just blindly installing everything you previously installed:
```
brew install $(/bin/ls /usr/local/Cellar)
```
I can't recommend this, because:
* it will likely error out and do no installs at all, if some packages no longer exist in Homebrew repos
* it will mark everything as being manually installed, so cleaning up packages that become unnecessary over time with `brew autoremove` will fail to do anything, and you'll be stuck carrying/updating unneeded packages forever

**Comment by nodesocket** [comment]
@SaschaKo77 Thanks for this comment, super helpful. I just got my MBP 16" M1 Max today and running `zsh` shell.

I've done:

```bash
# write a file with all existing brew packages out to Brewfile
brew bundle dump

# re-install Homebrew, will install into /opt/homebrew and use the arm version
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```

Any idea what I should set in `.zshrc` to use the new path for brew though? My brew is still using:

```shell
MacBook-Pro ➜  ~ which brew
/usr/local/bin/brew
```

Also, after I migrate all my packages from the `Brewfile` how can I completely uninstall the previous Intel version of Homebrew?

**Reply by nodesocket** [reply]
A helpful tip, to check if an executable is Intel or ARM you can use the `file` command like:

**Intel**
```bash
MacBook-Pro ➜  ~ file /usr/local/bin/git
/usr/local/bin/git: Mach-O 64-bit executable x86_64
```

**Apple Silicon (ARM)**
```bash
MacBook-Pro ➜  ~ file /opt/homebrew/bin/git
/opt/homebrew/bin/git: Mach-O 64-bit executable arm64

```

**Reply by gromgit** [reply]
> Any idea what I should set in `.zshrc` to use the new path for brew though?

Look for a line like this:
```
eval "$(/usr/local/bin/brew shellenv)"
```
and change it to invoke `brew` via the AS path:
```
eval "$(/opt/homebrew/bin/brew shellenv)"
```

**Reply by SaschaKo77** [reply]
Thanks for your kind words. In my .zshrc I added 
export PATH=/opt/homebrew/bin:$PATH

**Reply by dvessel** [reply]
To check all the executables from homebrew:
```for f in $(brew list); do echo $f && for e in $(brew unlink --dry-run $f | pcregrep -o1 "$(brew --prefix)/bin/(.*)"); do file `which $e`; done; done```

**Reply by gromgit** [reply]
A faster alternative that only uses standard system tools (and `brew`):
```
find $(brew --cellar) -path '*/*bin/*' -type f -exec file {} + | grep Mach-O
```

**Comment by nodesocket** [comment]
I've successfully migrated all my brew packages into` /opt/homebrew` and I'm ready to uninstall everything in the old `/usr/local/...`. What's the best practice way of uninstalling and removing all those packages?

The Homebrew uninstall script will uninstall everything I want to keep in `/opt/homebrew`:

https://raw.githubusercontent.com/Homebrew/install/HEAD/uninstall.sh

**Reply by carlocab** [reply]
Run the uninstall script under Rosetta.

**Reply by cho-m** [reply]
Another option is to use the uninstaller's [`--path=/usr/local`](https://github.com/Homebrew/install/blob/cc7fe005e7449d29a064d82736ad5419a4f9da23/uninstall.sh#L188) argument, which should override searching the default prefixes.

In case you use Rosetta, make sure either `/usr/local/bin/brew` is before `/opt/homebrew/bin/brew` on PATH, or remove both from PATH since the uninstall.sh currently [checks `brew --prefix` before the arch-default dir](https://github.com/Homebrew/install/blob/cc7fe005e7449d29a064d82736ad5419a4f9da23/uninstall.sh#L220-L224)

**Comment by alexreg** [comment]
Is there a recommended way to transfer over things installed by Homebrew formulas in `/usr/local`? For example (but not limited to), Python's or Node's packages. I realise that packages with native extensions may need recompilation, but this would be a good start at least.

**Reply by nodesocket** [reply]
Here is what I did, if I remember correctly.

- `brew bundle dump` to store all your currently installed packages in a file
- Reinstall brew using the script `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`
- Remove the old entry in your `.zshrc` or similar if running bash.
- Re-Add the entry in your `.zshrc` or similar 
- Run `which php` or some package you have installed and should show **NOT** in your path
- Reinstall all packages using your bundle dump as a reference manually. The path now should be `/opt/homebrew/bin/php`, etc.

```
MacBook-Pro ➜  ~ which php
/opt/homebrew/bin/php
```

- After that use the brew uninstaller script and set `--path=/usr/local` to remove all the old/legacy packages

**Reply by SMillerDev** [reply]
Any packages that weren't installed by brew can not be transferred by brew. I think both node and python have a way to query local packages and install that list again though

**Reply by gromgit** [reply]
If I remember correctly, you do this with Python:
```
/usr/local/bin/pip3 freeze > /tmp/python.lst
/opt/homebrew/bin/pip3 install --upgrade --force-reinstall -r /tmp/python.lst
```
Dunno about Node.

**Reply by SMillerDev** [reply]
maybe `npm ls --global --parseable --json` will work?

**Reply by alexreg** [reply]
Thanks for the advice, all. I'll give that a go.

**Reply by alexreg** [reply]
@SMillerDev How do I then install that list of packages?

**Reply by gromgit** [reply]
Not a Node user, but a quick bit of experimentation suggests the following for global modules:
```
/bin/ls /usr/local/lib/node_modules/ | while read m; do
  /opt/homebrew/bin/npm install --global "$m"
done
```

**Reply by alexreg** [reply]
@gromgit Thanks!

**Comment by tiptongrange** [comment]
After using brew bundle dump, is there any reason to not remove the Intel version first, before installing the M1 version, to remove any ambiguity of what's being removed?

**Reply by gromgit** [reply]
You could do that, or you could just tell the uninstaller which `HOMEBREW_PREFIX` to use, after you're sure the M1 version works for you:
```
curl -fsSL -o /tmp/uninstall.sh https://raw.githubusercontent.com/Homebrew/install/HEAD/uninstall.sh
/bin/bash /tmp/uninstall.sh --path=/usr/local
```

**Comment by gbelanger** [comment]
Thanks everyone for the tips. I went through and put this together for myself and my son to not spend the hour and a quarter doing this also.
```bash
## Migrating Homebrew from Intel to M1

## You can copy and paste into the Terminal

## Go to home directory
cd ;

## Create list of installed Intel packages
brew bundle dump ;

## Install new version of Homebrew
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)" ;

## Update the PATH to point to M1 version first
eval "$(/opt/homebrew/bin/brew shellenv)" ;
echo 'eval "$(/opt/homebrew/bin/brew shellenv)"' >> ~/.zshrc ;

## Migrate Intel packages (/usr/local/bin/brew) to M1 (/opt/Homebrew)
## You may need to enter your sysadmin password
brew bundle --file Brewfile ;

## Update packages
brew update ; 

## Upgrade packages
brew upgrade ;

## Clean up
brew cleanup ;

## Uninstall old Intel version

# Download uninstall script
curl -fsSL -o ./uninstall.sh https://raw.githubusercontent.com/Homebrew/install/HEAD/uninstall.sh ;

# Run the uninstall specifying the old path
sudo /bin/bash ./uninstall.sh --path=/usr/local ;
# Enter password
# Type Y to uninstall

# Delete uninstall script
rm uninstall.sh ;

# Final cleanup
brew cleanup ;
```

**Reply by nodesocket** [reply]
This is fantastic. We should add this script or an inspiration of it into brew or brew documentation.

**Reply by doraeimo** [reply]
I was using the migrated bash, located in `/usr/local/bin/bash`. The script installed Intel instead of ARM brew.
So First we'd better set the Shell open with `/bin/bash`

**Reply by Nauxscript** [reply]
```
brew bundle --file Brewfile ; 
```
this command will migrate the software installed by old brew to new brew? 
or just down the same software in new brew ?

**Reply by gromgit** [reply]
@Nauxscript It will _only_ reinstall the software listed in `Brewfile`. You are responsible for _manually_ migrating any required configuration and data files from `/usr/local` to `/opt/homebrew`.

**Reply by Nauxscript** [reply]
i have finished it and found that the softwar install in caskroom, like google-chrome, it can't  reinstall, and throw error like "there is already have a same app in my application folder", so i have to uninstall the software install in cask and reinstall them by hand.

**Reply by gromgit** [reply]
Yeah, that's a major difference between formulae and casks. Formulae are _completely_ stored under the Homebrew directory, which is different for different architectures.

Casks, on the other hand, actually _download and run the installer_ from the software's website, so the installs done by `brew bundle` would actually overwrite the original installation, hence the error messages you got.

You've already manually resolved the situation the hard way, but perhaps @Homebrew/cask folks might advise on the safety of "migrating" casks by simply copying the contents of `/usr/local/Caskroom` over to `/opt/homebrew/Caskroom`.

**Reply by SMillerDev** [reply]
Well, that won't actually install the applications so doesn't seem like a great solution.

**Reply by gromgit** [reply]
Well, the OP _did_ migrate the cask-installed apps to their new machine, hence the "already installed" errors from `brew bundle`. Assuming the migrated apps are universal, does copying the Caskroom directory over suffice?

**Reply by philippstorl** [reply]
Thanks. This helped me after some research.

**Reply by malikbenkirane** [reply]
It worked perfectly. I was struggling with podman formulae, and I'm done with it. Thank you so much!

**Reply by Gnosnay** [reply]
Thank you sooooooooooooooooo much!!! it helps a lot!

**Reply by dnszero** [reply]
I had the same problem and this worked great for me.

Thanks again for documenting this!

**Reply by petercunha** [reply]
Thanks man, you rock

**Comment by jchunkins** [comment]
I had the homebrew version of `bash` installed on my old mac, and did a Time Machine restore of a backup onto my new Apple Silicon mac. The new mac is running Mac OS Monterey. The version of `bash` I was using is not a universal binary, and only contains the x86_64 executable:

```console
$ lipo -info /usr/local/bin/bash
Non-fat file: /usr/local/bin/bash is architecture: x86_64
```

I've been using iTerm for my terminal, and despite the fact that I don't have iTerm set to `Open using Rosetta`, I would get the following from `uname` because of the version of `bash` I have set: 

```console
$ $SHELL --version
GNU bash, version 5.0.0(1)-release (x86_64-apple-darwin16.7.0)
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

$ uname -m
x86_64

$ uname -p
i386
```

Again, even if I were to use Apple's version of bash from `/bin/bash` the result would be the same:

```console
$ /bin/bash -c "uname -p"
i386

$ /bin/bash -c "uname -m"
x86_64
```

So as-is, if I were to attempt an install of homebrew I'd be installing the x86_64 version of homebrew, and not the Apple Silicon version.

So to ensure that the install would really give me the Apple Silicon version, I used the `arch -arm64` command to force Apple's version of bash (which is a universal binary) to run as arm64 architecture: 

```console
## prove that Apple's version of bash is really a universal binary
$ lipo -info /bin/bash
Architectures in the fat file: /bin/bash are: x86_64 arm64e

## prove that the arch -arm64 will really run as Apple Silicon
$ arch -arm64 uname -m
arm64

$ arch -arm64 uname -p
arm

$ arch -arm64 /bin/bash -c "uname -m"
arm64

$ arch -arm64 /bin/bash -c "uname -p"
arm

## perform the actual install
$ arch -arm64 /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```

Hope this helps.

**Reply by edwardchoijc** [reply]
I had the same problem as you, but I solved this with the following. Then you don't have to prefix every cmd with `arch -arm64`

```
ζ uname -m
x86_64

# Run your shell with the specific arch
ζ env /usr/bin/arch -arm64 /bin/zsh --login

ζ uname -m
arm64
```

What's more, you can switch back to Intel with param '-x86_64'

**Reply by Kryan90** [reply]
I was concerned I had just went through the whole process only to actually be reinstalling with x86_64 since my uname -m also returned x86_64, but checking with
```
brew config
...
HOMEBREW_PREFIX: /opt/homebrew
CPU: 10-core 64-bit arm_firestorm_icestorm
Rosetta 2: false
```
Seems to indicate it did still install the arm64 version

**Reply by jchunkins** [reply]
Once everything was installed, for me the easiest route was to set my bash profile to use these aliases:

```
alias ibrew='arch -x86_64 /usr/local/bin/brew'
alias mbrew='arch -arm64e /opt/homebrew/bin/brew'
```

One thing I did notice recently was the option for `-arm64e` instead of just `-arm64`, in the `arch` man page, so I used that since it calls out Apple Silicon explicitly. Not sure there's any difference between the two though.

At any rate, you can see the interesting differences between the two configurations: 

```
$ diff /tmp/a/ibrew.txt /tmp/a/mbrew.txt
1,7c1,7
< $ ibrew config
< HOMEBREW_VERSION: 3.4.11
< HOMEBREW_PREFIX: /usr/local
< Homebrew Ruby: 2.6.8 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8/bin/ruby
< CPU: 10-core 64-bit westmere
< macOS: 12.4-x86_64
< Rosetta 2: true
---
> $ mbrew config
> HOMEBREW_VERSION: 3.4.10
> HOMEBREW_PREFIX: /opt/homebrew
> Homebrew Ruby: 2.6.8 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
> CPU: 10-core 64-bit arm_firestorm_icestorm
> macOS: 12.4-arm64
> Rosetta 2: false
```
